### DISCUSSION TITLE
Scene Builder: Minimalist Mode and Nesting for Objects

### DISCUSSION THREAD

**Discussion by TerraPrimal** [4ian/GDevelop#2350] [Enhancement ideas (and their technical discussions)]
Greetings,

As my current project grows larger, my objects panel is starting to feel like a cluster f bomb ready to go off and I'd like to suggest a couple of things that may make things easier for larger projects.

**Minimalist Mode**
Most of my objects are named in a way that I don't need to see an icon/preview of them to know what they are. This option would remove the icon and cut the object size by around 50%.

![Alternate image text](https://i.gyazo.com/2a2c59b2473fc3d267ccbf966dbcc68c.png)
_Mockup showing the extra space saved_

---
**Nesting for Objects**
This option would help keep objects sorted and would purely be a visual feature. There are two ways that this could be done.
The first would be the same way we create groups in the event sheets where we just drag the objects into it and we can expand/close when required.

![Alternate image text](https://i.gyazo.com/09b841f6ed182e29599f61150f478505.png)
_I am sure it will look nicer in real life.._

The second way is how it's done in something like phpmyadmin where anything that has the same "prefix" is grouped.

![Alternate image text](https://i.gyazo.com/57e1f1780273249e9b436ca261fd0feb.png)
_Yes, I did start up XAMPP just for this example._

Thanks!

**Comment by 4ian** [comment]
Thanks for the suggestions, agree that the list can become cluttered/messy in large games. 

For the folder, have you tried to set tags to objects? Is this helping? My idea was to favor tagging instead of folders (because an object can have more than one tag, so it's more flexible in theory) and rely also on the search to speed up object navigation. Give me your thoughts.

**Comment by TerraPrimal** [comment]
Hi 4ian, thanks for the reply.

I do have tags set for my objects but it still feels messy when I don't have any tags enabled. If you're wanting to keep the tag system as the main way to filter objects, I'd highly suggest an inverse tag option.

This would allow the user to filter out objects with that tag but keep everything else visible.

**Comment by arthuro555** [comment]
Something @Silver-Streak suggested often too would be to use a tag grouping feature, where each tag would be showed as a folder containing all objects tagged with that tag. I personally really like the idea.

**Comment by Bouh** [comment]
> Something @Silver-Streak suggested often too would be to use a tag grouping feature, where each tag would be showed as a folder containing all objects tagged with that tag. I personally really like the idea.

The main issue with that is one object can be in two folders and more.
The tag system avoid to duplicate the objects.

Keep a clear list is important, group the object is cool, but have twice the object is bad.

**Comment by 4ian** [comment]
I've never thought about the inverted tags, though, it's an interesting idea!
It's true that for now tags need to put a bit of work to fully use them - it's way too easy to forget to add tags to your objects, while a folder is very "visible". Maybe we could have a way to select "untagged" objects, to find in a few seconds what objects you must organise?

**Comment by Silver-Streak** [comment]
> > Something @Silver-Streak suggested often too would be to use a tag grouping feature, where each tag would be showed as a folder containing all objects tagged with that tag. I personally really like the idea.
> 
> The main issue with that is one object can be in two folders and more.
> The tag system avoid to duplicate the objects.
> 
> Keep a clear list is important, group the object is cool, but have twice the object is bad.

I actually think objects showing up in multiple "tag folders" is totally okay, and in some cases super useful. E.g. I sometimes want to see all objects tagged "Enemies", but I may also want to see any items that use my "Floating" logic tagged "Floaters", which could include enemies. It's not a huge deal if the same object shows up in both "tag folders", so long as they don't appear in the same folder twice.

> I've never thought about the inverted tags, though, it's an interesting idea!
> It's true that for now tags need to put a bit of work to fully use them - it's way too easy to forget to add tags to your objects, while a folder is very "visible". Maybe we could have a way to select "untagged" objects, to find in a few seconds what objects you must organise?

Yeah, this matches up with my original though. "Untagged" or "Other" would show any objects without any tags, and make it super simple to manually tag and/or drag and drop them into a tag folder.

**Comment by Keevle** [comment]
i +1 folders for objs and "minimal mode",Having objs groups by folders isnt a bad idea,its universal from Art programs to code editors etc,i dont see why it would be a issue to do it for objs

**Comment by Bouh** [comment]
it is necessary to test new things, which make from an optimization point of view clearer and especially better ordered, the folders it is well but it is quickly galley when the file is duplicate.
It is up to the dev to put options or guards mad to avoid that the folders are a big mess.


If one day folder come,  I suggest to not duplicate the object, but make an reference, if an object is edited in an folder the original is updated too. This avoid the duplication in the project.

**Comment by Silver-Streak** [comment]
> If one day folder come, I suggest to not duplicate the object, but make an reference, if an object is edited in an folder the original is updated too. This avoid the duplication in the project.

Oh, yes, absolutely! To be clear, I meant that the item should *appear* twice, not that it's actually duplicated. Everything that is tagged should be a reference only to the original object, not actually a unique object itself. But that's me. 😄

**Comment by blurymind** [comment]
wouldnt object nesting, which is already on the trello board solve this?

**Comment by 4ian** [comment]
If I'm not mistaken, I think this is more about allowing a "container like" object that will be composed of other objects ("parent/child" objects).
Here the idea is more to group objects according to arbitrary categories (probably to differentiate for example UI objects from the rest..), in which case tags/folders are more adapted.

 > To be clear, I meant that the item should appear twice, not that it's actually duplicated. Everything that is tagged should be a reference only to the original object, not actually a unique object itself

Yes exactly, we display tags as folders (+ an "untagged" folder). If an object has multiple tags (which is very likely, especially on large scenes/games), it can be shown in multiple folders, but they are the same.
It still need a bit of work to go from a "dumb list" that can be filtered (by tags) to something more complex where we can show tags as folders (using tag names like "UI/Buttons", "UI/Minimap" to create "sub folders").
